examples/codes/external_code.ex.py

```diff
@@ -0,0 +1,428 @@
+    }
+
+    @property
+    def mappings(self) -> Dict:
+        """Code Mappings"""
+        return self._mappings
+
+    @classmethod
+    def from_defaults(cls) -> MappedParameterFrame:
+        """Setup from defaults"""
+        dd = {}
+        for _def in cls._defaults:
+            dd = {**dd, **asdict(_def)}
+        return super().from_defaults(dd)
+
+
+# %%
+class RunMode(BaseRunMode):
+    """
+    RunModes for external code
+    """
+
+    RUN = auto()
+    READ = auto()
+    MOCK = auto()
+
+
+# %% [markdown]
+# The `Setup` class pulls over the inputs from bluemira as described by the
+# mapping and creates the input file. The output of the run method returns
+# the commandline options list.
+
+# %%
+class Setup(CodesSetup):
+    """Setup task"""
+
+    params: ECParameterFrame
+
+    def __init__(self, params: ParameterFrame, problem_settings: Dict, infile: str):
+        super().__init__(params, PRG_NAME)
+        self.problem_settings = problem_settings
+        self.infile = infile
+
+    def update_inputs(self) -> Dict:
+        """Update inputs from bluemira"""
+        self.inputs = ECInputs()
+        self.options = ECOpts()
+        inp = self._get_new_inputs()
+        for thing in (self.problem_settings.items(), inp.items()):
+            for k, v in thing:
+                if k in self.options.__annotations__:
+                    if v:
+                        setattr(self.options, k, v)
+                    else:
+                        continue
+                elif v is not None:
+                    getattr(self.inputs, k)
+                    setattr(self.inputs, k, v)
+
+        return {k: v for k, v in asdict(self.inputs).items() if v}
+
+    def run(self) -> List:
+        """Run mode"""
+        inp = self.update_inputs()
+        with open(self.infile, "w") as _if:
+            for k, v in inp.items():
+                _if.write(f"{k}  {v}\n")
+        return self.options.to_list()
+
+
+# %% [markdown]
+# `Run` simply runs the code in a subprocess with the given options
+
+# %%
+class Run(CodesTask):
+    """Run task"""
+
+    def __init__(
+        self, params: ParameterFrame, infile: str, outfile: str, binary: List = BINARY
+    ):
+        super().__init__(params, PRG_NAME)
+        self.binary = binary
+        self.infile = infile
+        self.outfile = outfile
+
+    def run(self, options: List):
+        """Run mode"""
+        self._run_subprocess([*self.binary, *options, self.infile, self.outfile])
+
+
+# %% [markdown]
+# `Teardown` reads in a given output file or in the case of mock returns a known
+# value sending the new parameter values back to the `ParameterFrame`
+
+# %%
+class Teardown(CodesTeardown):
+    """Teardown task"""
+
+    def __init__(self, params: ParameterFrame, outfile: str):
+        super().__init__(params, PRG_NAME)
+        self.outfile = outfile
+
+    def _read_file(self):
+        out_params = {}
+        with open(self.outfile, "r") as of:
```

Comment:
To be consistent with my suggestion above.

```suggestion
        with open(self.outfile, "r") as output_file:
```